### PR TITLE
CommandControlVariables fixes

### PR DESCRIPTION
**Pull request by fmatthew5876**
When RPG_RT processes a command to set a range of variables to a
random value, it calls the RNG for each variable separately.

Fix: #2155 
Fix: #2156

@Ghabry can you please confirm it fixes the games in these issues? I don't have these japanese games working on my windows pc.

Confirmed this behavior in 2ke and 2k3e. Tested this with some example games.

### CODE REVIEW

**Comment by fmatthew5876** [comment]
jenkins test this please

**Comment by fmatthew5876** [comment]
Added the fix for #2156. I would prefer we store the correct map id internally for the consistency of our systems and add compatibility hacks like this at the edges.

Save games for this behavior are already correct as mentioned in #2156 comments.

**Comment by fmatthew5876** [comment]
I added a fix for direct and indirect variable lookups. Please review carefully and help test.

**Comment by Ghabry** [comment]
Does this have a speed impact on range operations? Older PRs had benchmarks.

File: src/game_interpreter.cpp [hunk 1]
```diff
@@ -1054,10 +1063,20 @@ bool Game_Interpreter::CommandControlVariables(RPG::EventCommand const& com) { /
...
-						value = character->GetMapId();
+						if (!Player::IsRPG2k()
+								|| event_id == Game_Character::CharPlayer
+								|| event_id == Game_Character::CharBoat
+								|| event_id == Game_Character::CharShip
+								|| event_id == Game_Character::CharAirship) {
+							value = character->GetMapId();
+						} else {
+							// This is an RPG_RT bug for 2k only. Requesting the map id of an event always returns 0.
+							value = 0;
```
Comment (Ghabry) [inline src/game_interpreter.cpp:1078]: This will "break" when you load a 2k3 save in 2k until you load a new map. But not worth to support this.

**Comment by fmatthew5876** [comment]
I haven't had time to measure. But my suspicions are:
* direct - most likely will, as you don't check resize the array each time. Doing 2 loops with a constant also avoids aliasing so the compiler can generate better code.
* indirect - likely will as well, but the cost of checking in Get(Get()) each time may dwarf it
* random - cost is probably overwhelmingly the rand call each time.

**Comment by Ghabry** [comment]
For getting japanese games to work you can (in a VM) set the Windows legacy codepage to Japanese. But this alters the behaviour for all non-unicode programs. btw there in the same dialog is now an option for "UTF-8" codepage like Linux does it but this also breaks tons of programs.

Or use https://github.com/xupefei/Locale-Emulator to run RPG_RT.

**Comment by fmatthew5876** [comment]
jenkins test this please

**Comment by fmatthew5876** [comment]
jenkins test this please

**Comment by Ghabry** [comment]
Can confirm that this fixes both issues.

**Comment by Ghabry** [comment]
Initialisation (I will always use this one)
```
@> Control Variables: [0011] = 2 
@> Control Variables: [0012] = 11 
@> Control Variables: [0013] = 5 
@> Control Variables: [0014] = 10 
@> Control Variables: [0015] = 20 
```

Indirect works properly but this fails for me:
```
@> Control Variables: [0011..0015] += Variable [0012]
```

**Comment by fmatthew5876** [comment]
Units tests and benchmarks added

```
2020-04-16 10:32:08
Running ./bench_variables
Run on (8 X 3504 MHz CPU s)
***WARNING*** Library was built as DEBUG. Timings may be affected.
---------------------------------------------------------------------------
Benchmark                                    Time           CPU Iterations
---------------------------------------------------------------------------
BM_VariableGet                               2 ns          2 ns  344615385
BM_VariableSet                               4 ns          4 ns  179200000
BM_VariableAdd                               4 ns          4 ns  172307692
BM_VariableSub                               4 ns          4 ns  179200000                                                                                                                                                                   BM_VariableMult                              4 ns          4 ns  165925926
BM_VariableDiv                               6 ns          5 ns  112000000
BM_VariableMod                               5 ns          5 ns  100000000
BM_VariableSetRange                        604 ns        614 ns    1120000
BM_VariableAddRange                        912 ns        921 ns     746667                                                                                                                                                                   BM_VariableSubRange                        607 ns        610 ns     896000
BM_VariableMultRange                       763 ns        767 ns     896000
BM_VariableDivRange                       2688 ns       2668 ns     263529
BM_VariableModRange                       2746 ns       2787 ns     263529
BM_VariableSetRangeVariable                618 ns        614 ns    1120000
BM_VariableSetRangeVariableIndirect       2569 ns       2567 ns     280000
BM_VariableSetRangeRandom                 8356 ns       8161 ns      74667
```

**Comment by fmatthew5876** [comment]
Just for fun if we change `SetRangeVariable` to do the naive thing and call `Get()` every time in a single loop.


```
BM_VariableSetRangeVariable               1645 ns       1650 ns     407273
```

**Comment by fmatthew5876** [comment]
It should be ok as long as we kick it out to a beta soon and let that have time to bake.

**Comment by fmatthew5876** [comment]
Confirmed RPG_RT (2ke here) this is just a dumb loop. Which explains the behavior.

```
char __fastcall sub_48CBE0(RPG_EventScriptLine *a1, int a2)
{
  RPG_EventScriptLine *v2; // ebx
  signed int v3; // eax
  int v4; // esi
  int v5; // eax
  signed int v6; // edi
  int v7; // ebp
  int v8; // esi
  signed int v9; // eax
  signed int v10; // eax
  signed int v11; // eax
  unsigned int v12; // eax
  bool v13; // cf
  int v14; // eax
  int v15; // eax
  int v16; // eax
  int v17; // edx
  RPG_Actor *v18; // eax
  int v19; // ecx
  RPG_Actor *v20; // eax
  int v21; // edx
  RPG_Actor *v22; // eax
  int v23; // ecx
  RPG_Actor *v24; // eax
  int v25; // edx
  RPG_Actor *v26; // eax
  int v27; // ecx
  int v28; // edx
  RPG_Actor *v29; // eax
  int v30; // ecx
  int v31; // edx
  RPG_Actor *v32; // eax
  int v33; // ecx
  RPG_Actor *v34; // eax
  int v35; // edx
  RPG_Actor *v36; // eax
  int v37; // ecx
  RPG_Actor *v38; // eax
  int v39; // edx
  RPG_Actor *v40; // eax
  int v41; // ecx
  RPG_Actor *v42; // eax
  int v43; // edx
  RPG_Actor *v44; // eax
  int v45; // ecx
  RPG_Actor *v46; // eax
  int v47; // edx
  RPG_Actor *v48; // eax
  int v49; // ecx
  RPG_Actor *v50; // eax
  int v51; // edx
  RPG_Actor *v52; // eax
  int v53; // ecx
  RPG_Actor *v54; // eax
  int v55; // edx
  RPG_Actor *v56; // eax
  int v57; // ecx
  RPG_Actor *v58; // eax
  int v59; // edx
  RPG_Actor *v60; // eax
  int v61; // ecx
  RPG_Actor *v62; // eax
  int v63; // edx
  RPG_Actor *v64; // eax
  int v65; // ecx
  RPG_Actor *v66; // eax
  int v67; // edx
  RPG_Actor *v68; // eax
  int v69; // ecx
  RPG_Actor *v70; // eax
  int v71; // edx
  RPG_Actor *v72; // eax
  int v73; // ecx
  RPG_Actor *v74; // eax
  int v75; // eax
  int v76; // edx
  RPG_Monster_Catalog *v77; // eax
  int v78; // ecx
  int v79; // edx
  RPG_Monster_Catalog *v80; // eax
  int v81; // ecx
  int v82; // edx
  RPG_Monster_Catalog *v83; // eax
  int v84; // ecx
  RPG_Monster *v85; // eax
  int v86; // edx
  RPG_Monster_Catalog *v87; // eax
  int v88; // ecx
  RPG_Monster *v89; // eax
  int v90; // edx
  RPG_Monster_Catalog *v91; // eax
  int v92; // ecx
  RPG_Monster *v93; // eax
  int v94; // edx
  RPG_Monster_Catalog *v95; // eax
  int v96; // ecx
  RPG_Actor *v97; // eax
  int v98; // edx
  RPG_Monster_Catalog *v99; // eax
  int v100; // ecx
  RPG_Monster *v101; // eax
  int v102; // edx
  RPG_Monster_Catalog *v103; // eax
  int v104; // ecx
  RPG_Monster *v105; // eax
  signed __int64 v106; // rax
  signed __int64 v107; // rax
  signed __int64 v108; // rax
  int a1a; // [esp+Ch] [ebp-28h]
  int v111; // [esp+10h] [ebp-24h]
  signed __int64 v112; // [esp+18h] [ebp-1Ch]
  RPG_Character *v113; // [esp+20h] [ebp-14h]

  v2 = a1;
  a1a = a2;
  if ( GetEventScriptLineParameter(a1, 0) == 2 )
  {
    v3 = GetEventScriptLineParameter(v2, 1);
    v4 = GetVariable(*RPG::system, v3);
  }
  else
  {
    v4 = GetEventScriptLineParameter(v2, 1);
  }
  if ( GetEventScriptLineParameter(v2, 0) == 1 )
    v5 = GetEventScriptLineParameter(v2, 2);
  else
    v5 = v4;
  v6 = v4;
  if ( v5 >= v4 )
  {
    v7 = v5 - v4 + 1;
    do
    {
      v8 = 0;
      switch ( GetEventScriptLineParameter(v2, 4) )
      {
        case 0:
          v8 = GetEventScriptLineParameter(v2, 5);
          break;
        case 1:
          v9 = GetEventScriptLineParameter(v2, 5);
          v8 = GetVariable(*RPG::system, v9);
          break;
        case 2:
          v10 = GetEventScriptLineParameter(v2, 5);
          v11 = GetVariable(*RPG::system, v10);
          v8 = GetVariable(*RPG::system, v11);
          break;
        case 3:
          v8 = GetEventScriptLineParameter(v2, 5);
          v111 = GetEventScriptLineParameter(v2, 6);
          if ( v8 <= v111 )
          {
            if ( v8 < v111 )
              v8 += RPG_RandExclusive(v111 - v8 + 1);
          }
          else
          {
            v8 = v111 + RPG_RandExclusive(v8 - v111 + 1);
          }
          break;
        case 4:
          v12 = GetEventScriptLineParameter(v2, 6);
          v13 = v12 < 1;
          v14 = v12 - 1;
          if ( v13 )
          {
            v15 = GetEventScriptLineParameter(v2, 5);
            v8 = GetItemCountInInventory(*RPG::Inventory, v15);
          }
          else if ( !v14 )
          {
            v16 = GetEventScriptLineParameter(v2, 5);
            v8 = GetItemCountEquipped(*RPG::Inventory, v16);
          }
          break;
        case 5:
          switch ( GetEventScriptLineParameter(v2, 6) )
          {
            case 0:
              v17 = GetEventScriptLineParameter(v2, 5);
              v18 = *RPG::actors;
              v20 = LookupActor(v19, v17);
              v8 = GetActorLevel(v20);
              break;
            case 1:
              v21 = GetEventScriptLineParameter(v2, 5);
              v22 = *RPG::actors;
              v24 = LookupActor(v23, v21);
              v8 = GetActorExp(v24);
              break;
            case 2:
              v25 = GetEventScriptLineParameter(v2, 5);
              v26 = *RPG::actors;
              v8 = LookupActor(v27, v25)->battler.hp;
              break;
            case 3:
              v28 = GetEventScriptLineParameter(v2, 5);
              v29 = *RPG::actors;
              v8 = LookupActor(v30, v28)->battler.mp;
              break;
            case 4:
              v31 = GetEventScriptLineParameter(v2, 5);
              v32 = *RPG::actors;
              v34 = LookupActor(v33, v31);
              v8 = (v34->battler.vTable->GetMaxHp)();
              break;
            case 5:
              v35 = GetEventScriptLineParameter(v2, 5);
              v36 = *RPG::actors;
              v38 = LookupActor(v37, v35);
              v8 = (v38->battler.vTable->GetMaxMp)();
              break;
            case 6:
              v39 = GetEventScriptLineParameter(v2, 5);
              v40 = *RPG::actors;
              v42 = LookupActor(v41, v39);
              v8 = GetBattlerAtk(&v42->battler);
              break;
            case 7:
              v43 = GetEventScriptLineParameter(v2, 5);
              v44 = *RPG::actors;
              v46 = LookupActor(v45, v43);
              v8 = GetActorDef(v46);
              break;
            case 8:
              v47 = GetEventScriptLineParameter(v2, 5);
              v48 = *RPG::actors;
              v50 = LookupActor(v49, v47);
              v8 = GetActorInt(&v50->battler);
              break;
            case 9:
              v51 = GetEventScriptLineParameter(v2, 5);
              v52 = *RPG::actors;
              v54 = LookupActor(v53, v51);
              v8 = GetBattlerAgi(&v54->battler);
              break;
            case 10:
              v55 = GetEventScriptLineParameter(v2, 5);
              v56 = *RPG::actors;
              v58 = LookupActor(v57, v55);
              v8 = GetActorEquipment(v58, 0);
              break;
            case 11:
              v59 = GetEventScriptLineParameter(v2, 5);
              v60 = *RPG::actors;
              v62 = LookupActor(v61, v59);
              v8 = GetActorEquipment(v62, 1);
              break;
            case 12:
              v63 = GetEventScriptLineParameter(v2, 5);
              v64 = *RPG::actors;
              v66 = LookupActor(v65, v63);
              v8 = GetActorEquipment(v66, 2);
              break;
            case 13:
              v67 = GetEventScriptLineParameter(v2, 5);
              v68 = *RPG::actors;
              v70 = LookupActor(v69, v67);
              v8 = GetActorEquipment(v70, 3);
              break;
            case 14:
              v71 = GetEventScriptLineParameter(v2, 5);
              v72 = *RPG::actors;
              v74 = LookupActor(v73, v71);
              v8 = GetActorEquipment(v74, 4);
              break;
            default:
              goto LABEL_64;
          }
          break;
        case 6:
          v75 = GetEventScriptLineParameter(v2, 5);
          v113 = GetEventById(a1a, v75);
          switch ( GetEventScriptLineParameter(v2, 6) )
          {
            case 0:
              v8 = v113->mapId;
              break;
            case 1:
              v8 = v113->x;
              break;
            case 2:
              v8 = v113->y;
              break;
            case 3:
              v8 = dword_4A090C[v113->facing];
              break;
            case 4:
              v8 = GetCharacterScreenX(v113);
              break;
            case 5:
              v8 = GetCharacterScreenY(v113);
              break;
            default:
              goto LABEL_64;
          }
          break;
        case 7:
          switch ( GetEventScriptLineParameter(v2, 5) )
          {
            case 0:
              v8 = (*RPG::Inventory)->goldAmount;
              break;
            case 1:
              v8 = GetTimerSeconds(*RPG::Inventory);
              break;
            case 2:
              v8 = (*RPG::Inventory)->party.size;
              break;
            case 3:
              v8 = (*RPG::system)->saveCount;
              break;
            case 4:
              v8 = (*RPG::Inventory)->numBattles;
              break;
            case 5:
              v8 = (*RPG::Inventory)->numVictories;
              break;
            case 6:
              v8 = (*RPG::Inventory)->numDefeats;
              break;
            case 7:
              v8 = (*RPG::Inventory)->numEscapes;
              break;
            case 8:
              v8 = sub_45CCC8();
              break;
            default:
              goto LABEL_64;
          }
          break;
        case 8:
          switch ( GetEventScriptLineParameter(v2, 6) )
          {
            case 0:
              v76 = GetEventScriptLineParameter(v2, 5);
              v77 = *RPG::monsters;
              v8 = LookupMonster(v78, v76)->battler.hp;
              break;
            case 1:
              v79 = GetEventScriptLineParameter(v2, 5);
              v80 = *RPG::monsters;
              v8 = LookupMonster(v81, v79)->battler.mp;
              break;
            case 2:
              v82 = GetEventScriptLineParameter(v2, 5);
              v83 = *RPG::monsters;
              v85 = LookupMonster(v84, v82);
              v8 = (v85->battler.vTable->GetMaxHp)();
              break;
            case 3:
              v86 = GetEventScriptLineParameter(v2, 5);
              v87 = *RPG::monsters;
              v89 = LookupMonster(v88, v86);
              v8 = (v89->battler.vTable->GetMaxMp)();
              break;
            case 4:
              v90 = GetEventScriptLineParameter(v2, 5);
              v91 = *RPG::monsters;
              v93 = LookupMonster(v92, v90);
              v8 = GetBattlerAtk(&v93->battler);
              break;
            case 5:
              v94 = GetEventScriptLineParameter(v2, 5);
              v95 = *RPG::monsters;
              v97 = LookupMonster(v96, v94);
              v8 = GetActorDef(v97);
              break;
            case 6:
              v98 = GetEventScriptLineParameter(v2, 5);
              v99 = *RPG::monsters;
              v101 = LookupMonster(v100, v98);
              v8 = GetActorInt(&v101->battler);
              break;
            case 7:
              v102 = GetEventScriptLineParameter(v2, 5);
              v103 = *RPG::monsters;
              v105 = LookupMonster(v104, v102);
              v8 = GetBattlerAgi(&v105->battler);
              break;
            default:
              goto LABEL_64;
          }
          break;
        default:
          break;
      }
LABEL_64:
      v112 = GetVariable(*RPG::system, v6);
      if ( GetEventScriptLineParameter(v2, 3) >= 4 && !v8 )
        v8 = 1;
      switch ( GetEventScriptLineParameter(v2, 3) )
      {
        case 0:
          v112 = v8;
          break;
        case 1:
          v112 += v8;
          break;
        case 2:
          v112 -= v8;
          break;
        case 3:
          LODWORD(v106) = Multiply(v112, v8, v8 >> 31);
          v112 = v106;
          break;
        case 4:
          LODWORD(v107) = sub_405D22(v112, v8, v8 >> 31);
          v112 = v107;
          break;
        case 5:
          LODWORD(v108) = sub_405E15(v112, v8, v8 >> 31);
          v112 = v108;
          break;
        default:
          break;
      }
      if ( HIDWORD(v112) == -1 )
      {
        if ( v112 >= 0xFFF0BDC1 )
          goto LABEL_79;
      }
      else if ( SHIDWORD(v112) >= -1 )
      {
LABEL_79:
        if ( v112 <= 999999 )
          SetVariable(*RPG::system, v6, v112);
        else
          SetVariable(*RPG::system, v6, 999999);
        goto LABEL_82;
      }
      SetVariable(*RPG::system, v6, -999999);
LABEL_82:
      ++v6;
      --v7;
    }
    while ( v7 );
  }
  return MaybeRefreshMapEvents((*RPG::mapEvents)[2].vTable);
}
```

**Comment by Ghabry** [comment]
Okay so this is as expected. thanks for confirming.

**Pull request merged**